### PR TITLE
Satellite accepts manipulated paychecks V3-1023

### PR DESCRIPTION
**Pull request by aligeti**
https://storjlabs.atlassian.net/browse/V3-1022

### CODE REVIEW

File: satellite/satellitedb/bandwidthagreement.go [hunk 1]
```diff
@@ -8,20 +8,35 @@ import (
...
-		dbx.Bwagreement_Data(agreement.Agreement),
-	)
-	return err
+func (b *bandwidthagreement) CreateAgreement(ctx context.Context, serialNum string, agreement bwagreement.Agreement) error {
+	tx, err := b.db.Open(ctx)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	_, err = tx.Get_Bwagreement_By_Serialnum(ctx, dbx.Bwagreement_Serialnum(serialNum))
```
Comment (cam-a) [inline satellite/satellitedb/bandwidthagreement.go]: I don't think we need the serialNum query and transaction. The serial number field is defined as `unique`, so an error is returned if we attempt to enter an agreement with a pre-existing serial number

File: pkg/bwagreement/server.go [hunk 2]
```diff
@@ -52,7 +52,7 @@ func NewServer(db DB, logger *zap.Logger, pkey crypto.PublicKey) *Server {
 }
 
 // BandwidthAgreements receives and stores bandwidth agreements from storage nodes
-func (s *Server) BandwidthAgreements(ctx context.Context, req *pb.RenterBandwidthAllocation) (reply *pb.AgreementsSummary, err error) {
+func (s *Server) BandwidthAgreements(ctx context.Context, ba *pb.RenterBandwidthAllocation) (reply *pb.AgreementsSummary, err error) {
```
[the remark this answers is not in the document]
Comment (littleskunk) [reply pkg/bwagreement/server.go]: Sure we also have a problem on the storage node side. At the moment it is possible to let the storage node work for free. That would require a manipulated uplink. For the storage node alpha release we will run the uplinks. So lets ignore that for the moment and fix it with another PR.

File: pkg/bwagreement/server.go [hunk 3]
```diff
@@ -61,13 +61,31 @@ func (s *Server) BandwidthAgreements(ctx context.Context, req *pb.RenterBandwidt
 		Status: pb.AgreementsSummary_FAIL,
 	}
 
-	if err = s.verifySignature(ctx, req); err != nil {
+	if err = s.verifySignature(ctx, ba); err != nil {
```
[the remark this answers is not in the document]
Comment (littleskunk) [reply pkg/bwagreement/server.go]: That should't be a problem here. The satellite will write the uplink pubkey into his PayerBandwidthAgreement. That way the Payer and RenterBandwidthAgreements can't be manipulated.

File: pkg/accounting/tally/tally_test.go [hunk 4]
```diff
@@ -71,7 +71,7 @@ func TestQueryWithBw(t *testing.T) {
 	assert.NoError(t, err)
 	//save to db
 
-	err = bwDb.CreateAgreement(ctx, bwagreement.Agreement{Signature: rba.GetSignature(), Agreement: rba.GetData()})
+	err = bwDb.CreateAgreement(ctx, "SerialNumber", bwagreement.Agreement{Signature: rba.GetSignature(), Agreement: rba.GetData()})
```
[the remark this answers is not in the document]
Comment (littleskunk) [reply pkg/accounting/tally/tally_test.go]: Not possible. The bwagreement contains the storagenodeID. Duplicating it to dozens of rivals only works if the rivals don't verify the storagenodeID. Even without any duplicates they would be able to submit it but the satellite will pay the storagenodeID that was written into the agreement. The satellite will not care who submitted it.

**Pull request merged**